### ISSUE TITLE
Trying to understand the differences between Pytext, Allennlp and Torchtext?

### ISSUE THREAD

**Issue by hohoCode** [facebookresearch/pytext#110]
Thanks for the opensource. I am just trying to understand the differences between Pytext and Allennlp? Since both say they can help manage NLP experiments. Or what do you think on Pytext's advantage over Allennlp? THANKS!

**Comment by matthew-z** [comment]
In my opinion 
###  Pytext

* based on torchtext;
* supports distributed training;
* supports multi-task training;
* currently only supports some tasks.


### Allennlp
* Provides a lot of handful modules like Seq2VecEncoder, Seq2SeqEncoder.
* Provides a lot of handful utilities like  masked_softmax, masked_log_softmax .....  (You can get a lot of mask-related helper functions in AllenNLP, but there is no such thing in pytext).
* A lot of embeddings and tokenization are out of the box, such as ELMo, BERT.
* configs are based on jsonnet, which is much more powerful than json (pytext uses json I think).
* Provides model/data_reader for a lot of popular NLP tasks.
* No distributed training.
* Multi-task training is not very straightforward in some cases.

(I just read the pytext doc, so I might be wrong)

**Comment by ahhegazy** [comment]
**Torchtext** Provides a set of data-abstractions that helps reading and processing raw text data into PyTorch tensors, at the moment we use Torchtext in PyText for training-time data reading and preprocessing.

**AllenNLP** is a great NLP modeling library that is aimed at providing reference implementations and prebuilt state-of-the-art models, and make it easy to iterate on and research with models for different NLP tasks.

**PyText**
We've built PyText to be a rich NLP modeling library (along the lines of AllenNLP) but with production capabilities baked in the design from day 1. Examples are:

- We provide interfaces to make sure data preprocessing can be consistent between training and runtime 
- The model interfaces are compatible with ONNX and torch.jit 
- A core goal for us in the next few month is to be able to run models trained in PyText on mobile.

Among other differences like supporting distributed training and multi-task training.

That being said, so far our library of models has been mostly influenced by our current production use-cases, we are actively working on enriching this library with more models and tasks while keeping production capabilities and inference speed in mind.

**Comment by whiletruelearn** [comment]
@ahhegazy Thanks for the comparison. Is there plan to bring `fasttext` also under `PyText`

**Comment by m3rlin45** [comment]
If you mean in the sense of unifying the projects, no, there are no plans. If you mean supporting fasttext models in PyText, we are evaluating that, as it is definitely a feature request we've heard.

**Comment by maxdel** [comment]
Thanks for the comparison, are there any plans for integration with [fairseq-py](https://github.com/pytorch/fairseq) and/or [translate-py](https://github.com/pytorch/translate)?

**Comment by ahhegazy** [comment]
Yes, we are currently working on that!
